File: src/handlers/katex_assets.rs
use once_cell::sync::Lazy;
use std::env;
use std::sync::atomic::{AtomicBool, Ordering};

// Track if KaTeX was used in the current page (thread-local)
thread_local! {
    static KATEX_USED: AtomicBool = const { AtomicBool::new(false) };
}

// Track if we've printed the KaTeX message (global, one-time)
static MESSAGE_PRINTED: Lazy<AtomicBool> = Lazy::new(|| AtomicBool::new(false));

pub fn mark_katex_used() {
    KATEX_USED.with(|used| used.store(true, Ordering::Relaxed));
}

pub fn was_katex_used() -> bool {
    KATEX_USED.with(|used| used.load(Ordering::Relaxed))
}

pub fn reset_katex_flag() {
    KATEX_USED.with(|used| used.store(false, Ordering::Relaxed));
}

pub fn print_katex_message() {
    if !MESSAGE_PRINTED.swap(true, Ordering::Relaxed) {
        println!("  📐 KaTeX CSS will be injected (using CDN)");
    }
}

pub fn is_katex_injection_disabled() -> bool {
    env::var("SIMPLE_DISABLE_KATEX_CSS").is_ok()
}

pub fn get_katex_css_tag() -> &'static str {
    r#"<!-- KaTeX CSS (auto-injected from CDN) -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">"#
}

File: src/handlers/katex_assets.cpp
#include "handlers/katex_assets.hpp"
#include <atomic>
#include <cstdlib>
#include <iostream>

namespace simple::handlers {

thread_local std::atomic<bool> katex_used{false};
static std::atomic<bool> message_printed{false};

auto mark_katex_used() -> void {
  katex_used.store(true, std::memory_order_relaxed);
}

auto was_katex_used() -> bool {
  return katex_used.load(std::memory_order_relaxed);
}

auto reset_katex_flag() -> void {
  katex_used.store(false, std::memory_order_relaxed);
}

auto print_katex_message() -> void {
  bool expected = false;
  if (message_printed.compare_exchange_strong(expected, true,
                                              std::memory_order_relaxed)) {
    std::cout << "  📐 KaTeX CSS will be injected (using CDN)\n";
  }
}

auto is_katex_injection_disabled() -> bool {
  return std::getenv("SIMPLE_DISABLE_KATEX_CSS") != nullptr;
}

auto get_katex_css_tag() -> std::string_view {
  return R"(<!-- KaTeX CSS (auto-injected from CDN) -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.16.11/dist/katex.min.css" integrity="sha384-nB0miv6/jRmo5UMMR1wu3Gz6NLsoTkbqJghGIsx//Rlm+ZU03BU6SQNC66uf4l5+" crossorigin="anonymous">)";
}

} // namespace simple::handlers
